src/KOKKOS/nbin_ssa_kokkos.cpp

```diff
@@ -0,0 +1,311 @@
+{
+  if (mbins > (int) k_bins.h_view.dimension_0()) {
+    k_bins = DAT::tdual_int_2d("NBinSSAKokkos::bins",mbins,atoms_per_bin);
+    bins = k_bins.view<DeviceType>();
+
+    k_bincount = DAT::tdual_int_1d("NBinSSAKokkos::bincount",mbins);
+    bincount = k_bincount.view<DeviceType>();
+  }
+
+  ghosts_per_gbin = atom->nghost / 7; // estimate needed size
+
+  if (ghosts_per_gbin > (int) k_gbins.h_view.dimension_1()) {
+    k_gbins = DAT::tdual_int_2d("NBinSSAKokkos::gbins",8,ghosts_per_gbin);
+    gbins = k_gbins.view<DeviceType>();
+  }
+
+  // Clear the local bin extent bounding box.
+  h_lbinxlo() = mbinx - 1; // Safe to = stencil->sx + 1
+  h_lbinylo() = mbiny - 1; // Safe to = stencil->sy + 1
+  h_lbinzlo() = mbinz - 1; // Safe to = stencil->sz + 1
+  h_lbinxhi() = 0; // Safe to = mbinx - stencil->sx - 1
+  h_lbinyhi() = 0; // Safe to = mbiny - stencil->sy - 1
+  h_lbinzhi() = 0; // Safe to = mbinz - stencil->sz - 1
+  deep_copy(d_lbinxlo, h_lbinxlo);
+  deep_copy(d_lbinylo, h_lbinylo);
+  deep_copy(d_lbinzlo, h_lbinzlo);
+  deep_copy(d_lbinxhi, h_lbinxhi);
+  deep_copy(d_lbinyhi, h_lbinyhi);
+  deep_copy(d_lbinzhi, h_lbinzhi);
+}
+
+/* ----------------------------------------------------------------------
+   bin owned and ghost atoms for the Shardlow Splitting Algorithm (SSA)
+   local atoms are in distinct bins (binhead[]) from the ghosts
+   ghost atoms are "binned" in gairhead_ssa[] instead
+     ghosts which are not in an Active Interaction Region (AIR) are skipped
+------------------------------------------------------------------------- */
+
+template<class DeviceType>
+void NBinSSAKokkos<DeviceType>::bin_atoms()
+{
+  last_bin = update->ntimestep;
+
+  int nlocal = atom->nlocal;
+  int nghost = atom->nghost;
+  int nall = nlocal + nghost;
+
+  atomKK->sync(ExecutionSpaceFromDevice<DeviceType>::space,X_MASK);
+  x = atomKK->k_x.view<DeviceType>();
+
+  sublo_[0] = domain->sublo[0];
+  sublo_[1] = domain->sublo[1];
+  sublo_[2] = domain->sublo[2];
+  subhi_[0] = domain->subhi[0];
+  subhi_[1] = domain->subhi[1];
+  subhi_[2] = domain->subhi[2];
+
+  bboxlo_[0] = bboxlo[0]; bboxlo_[1] = bboxlo[1]; bboxlo_[2] = bboxlo[2];
+  bboxhi_[0] = bboxhi[0]; bboxhi_[1] = bboxhi[1]; bboxhi_[2] = bboxhi[2];
+
+  k_binID = DAT::tdual_int_1d("NBinSSAKokkos::binID",nall);
+  binID = k_binID.view<DeviceType>();
+
+  // find each local atom's binID
+  {
+    atoms_per_bin = 0;
+    NPairSSAKokkosBinIDAtomsFunctor<DeviceType> f(*this);
+    Kokkos::parallel_reduce(nlocal, f, atoms_per_bin);
+  }
+  deep_copy(h_lbinxlo, d_lbinxlo);
+  deep_copy(h_lbinylo, d_lbinylo);
+  deep_copy(h_lbinzlo, d_lbinzlo);
+  deep_copy(h_lbinxhi, d_lbinxhi);
+  deep_copy(h_lbinyhi, d_lbinyhi);
+  deep_copy(h_lbinzhi, d_lbinzhi);
+
+  // find each ghost's binID (AIR number)
+  {
+    for (int i = 0; i < 8; i++) k_gbincount.h_view(i) = 0;
+    k_gbincount.modify<LMPHostType>();
+    k_gbincount.sync<DeviceType>();
+    DeviceType::fence(); // FIXME?
```

Comment:
@stanmoore1 Does this modify(), sync(), fence() sequence (and the one below on line 170 with the FIXME comment make sense here?
I was just doing "monkey see monkey do" coding when this went in, and I wasn't sure if it was correct or not.
It looks correct to me now, but I put "FIXME?" there for some reason.

Reply:
The sync/modify is correct. The fence isn't necessary and can be removed. I removed all the fences out of the fixes and pair styles but didn't modify other styles.

Reply:
@timattox There are only a few places in LAMMPS where a thread fence is needed, like in verlet_kokkos

Reply:
But DeviceType::fence() occurs in a bunch of .cpp files in the KOKKOS directory. Yes, verlet_kokkos.cpp is the only one with Kokkos::fence(), but I don't see how that is relevant?
(Tim is confused... and needs to eat lunch before trying to understand...)

Reply:
Okay, after talking with @stanmoore1, I'll leave this alone, and he will deal with the removal of extraneous DeviceType::fence() calls as he has time to test and verify their removal.